doc/api/permissions.md

```diff
@@ -506,6 +510,45 @@ process.permission.has('fs.read'); // true
 process.permission.has('fs.read', '/home/rafaelgss/protected-folder'); // false
 ```
 
+#### Environment Permissions
+
+To configure permission to access the environment variables via `process.env`,
+use the [`--allow-env`][] flag:
+
+```console
+$ node --experimental-permission --allow-fs-read=* --allow-env=* index.js
+```
+
+The valid arguments for the flag are:
+
+* Strings delimited by comma (`,`).
+* `*` - Used as a wildcard. To allow access to all environment variables, simply
+  use `--allow-env=*`.
+  * Use a string ended with `*` to allow access to variables that match the
+    specified prefix. For example, `--allow-env=NODE_*` will allow access to all
+    variables that start with `NODE_`.
+  * If `*` is not used as the last character of a string, such as `N*DE`,`*ODE`,
+    or `*ODE*`, the string is ignored.
+* `-` - Used to deny access. By prefixing variables with `-`, you can restrict
```

Comment:
It may be a bit confusing, but considering use cases of environment permissions, I think it's worth adding. That's because the most common scenarios are those where only a few variables are denied, while access to all the other environment variables is allowed.

Whether to use the `-` syntax requires further discussion. I think that it's more convenient compared to introducing a separate option like `--env-deny`.

Reply:
"-" is a valid character for environment variable names though.
e.g. on Windows
```console
C:\Users\rlau>echo %-TESTVAR%
%-TESTVAR%

C:\Users\rlau>set -TESTVAR=value

C:\Users\rlau>echo %-TESTVAR%
value
```

On Linux bash doesn't appear to let you set e.g. `-TESTVAR`, but you can do it via Node.js and spawn a child process which will inherit it:
```console
$ ./node
Welcome to Node.js v21.0.0-pre.
Type ".help" for more information.
> process.env['-TESTVAR']='value'
'value'
> child_process.execSync('node -p process.env').toString()
'{\n' +
...
  "  '-TESTVAR': 'value',\n" +
...
  '}\n'
>

Reply:
I think you were quering about blocking variables like `-TESTVAR`.

The `-` is a syntax of the flag used to deny access, and this parses the string following the first `-` as variable names. So the `-TESTVAR` can be denied also as below.

```js
console.log(process.env.USER);
process.env['-TESTVAR']='value'
```
```console
$ node --experimental-permission --allow-fs-read=* --allow-env=*,--TESTVAR ./test.js
daeyeon
/home/daeyeon/code/nodeland/daeyeon.node/test.js:2
process.env['-TESTVAR']='value'
                       ^
Error: Access to this API has been restricted
    ....
    at node:internal/main/run_main_module:23:47 {
  code: 'ERR_ACCESS_DENIED',
  permission: 'Environment',
  resource: '-TESTVAR'
}
```

Reply:
@daeyeon My point is that both `TESTVAR` and `-TESTVAR` are valid names for environment variables. It would appear to be impossible to allow just `-TESTVAR` without using wildcards?

Reply:
> That's because the most common scenarios are those where only a few variables are denied, while access to all the other environment variables is allowed.

@daeyeon Could you give an example? If denying access to a few specific variables is the most common scenario, and if those have to be specified when starting the Node.js process, why can't the user just unset those specific environment variables when starting the Node.js process?

Reply:
I echo that question; I've never run into that scenario myself - anything in the env is leakable, so anything not leakable can't be exposed to the app in the first place.

Reply:
@richardlau Thank you for awakening me. I see now, so that means wildcard inputs like `--allow-env=*` or `--allow-env=NODE_*` are also not possible. Therefore, to grant full access to environment variables using a flag, I guess that the only option available is to use `--allow-env` with no reference.

Reply:
@tniessen 

> why can't the user just unset those specific environment variables when starting the Node.js process?

My assumptions for this feature are that users may not be familiar with all the ways to set/unset environment variables in different Node.js execution environments, and that Node.js does not run within a specific sandbox. Users rely on Node.js at the application level to control sensitive environment variables.

For instance, when running a Node.js app without knowing what environment variable accesses are required by the app, users could rely on: a) permissions specified in any of developer's manifest policy file (--allow-env), b) manually granting permissions using prompts if the manifest policy file is absent, or c) granting permissions to all environment variables except sensitive information (--deny-env).

Initially, I considered c) to be the most common scenario when a) and b) are not present, but I now realize that was somewhat narrow.

Reply:
If they're not familiar with those ways, they certainly won't be familiar with `--allow-env` - a near-universal shell standard is far more likely to be known than a node-specific argument.

Reply:
Yes, I seem to have considered uncommon scenarios. The deny option isn't currently available in other permissions, so it's definitely inconsistent. I'll withdraw the proposal for the deny option and change this PR to only support allowing environment variables.

Reply:
Given that this is a 1500 line change, I think the main question is whether this is a useful feature that is worth the complexity. As @ljharb said, users can already easily unset sensitive variables through other means.

Reply:
Could I know how easy it is? Maybe the "easily" is related to the number of variables that users want to unset?

Reply:
Since the difficulty of, say, unsettling 10 vs 1 is the same with any approach, i don’t think that really impacts it. No matter how many you want to unset, iow, it’d still be better and easier to use the universally understood approach than a redundant node-specific one.

Reply:
Thanks for the opinion. I still have a different perspective.

Let's say there is a Node.js app reading `process.env.HOME` variable (not using `node:os`).

1. Users want to expose `process.env.HOME` only. They can use `--allow-env=HOME` flag to do it. In this case, guiding them to unset all remaining variables to prevent potential leaks would be inconvenient. Plus, since deno already has this feature, this inconvenience would be compared to it.

2. Once enabling Node.js permission feature, users can track attempts by the app to access user's other environment variable without having to examine the entire app code (including node_modules). This can be done by either throwing an ERR_ACCESS_DENIED or providing promptying access, similar to deno.

While 1 may vary depending on personal perspectives, I doubt there is a universal way to achieve 2.

Reply:
For 1, `env -i HOME=$HOME node whatever` i think would do it?

There's noth

Reply:
> Plus, since deno already has this feature, this inconvenience would be compared to it.

Deno was designed to facilitate a somewhat useful permission system from day one. Node.js wasn't. I don't think adding questionable features in an attempt to catch up with Deno should be a goal of the Node.js project.

> Once enabling Node.js permission feature, users can track attempts by the app to access user's other environment variable without having to examine the entire app code

If tracking access to individual environment variables was desirable, I'm sure a modern Linux kernel would support that. It doesn't, because it's effectively just a list of named arguments that are passed to a process. Don't want the process to see it, don't pass it.

---

There is still no real roadmap for the permission model (see https://github.com/nodejs/security-wg/issues/898), so I don't know what this is supposed to achieve. If integration with package managers in development environments is the main goal (because production environments are clearly not the main goal), wouldn't it be easy for the package manager to only selectively pass environment variables to child processes?

Reply:
> For 1, env -i HOME=$HOME node whatever i think would do it?

Well... Windows developers will need different a guide, and some would prefer Node.js flags (as a cross platform way) when having more names to pass.

Reply:
> There is still no real roadmap for the permission model (see https://github.com/nodejs/security-wg/issues/898), so I don't know what this is supposed to achieve. If integration with package managers in development environments is the main goal (because production environments are clearly not the main goal), wouldn't it be easy for the package manager to only selectively pass environment variables to child processes?

It's true that we can rely on host system mechanisms to manage environment variables as well as other operations (fs, net, and so on). Doubtful about Node.js built-in permissions system itself? I haven't caught every discussion, so it's hard to say. I can find that environment variable permission is on the roadmap, and I just want to move the action item forward with this PR and discussions.

Reply:
Created a new [thread](https://github.com/nodejs/security-wg/issues/993) to separate implementation and further discussions.